GPS: Configure COEX0 and MAGPIO for Thingy:91 in driver and samples

jtguggedal opened the pull request:
- Add COEX0 and MAGPIO handling for Thingy:91 to the nRF9160 driver by default
- Add COEX0 and MAGPIO commands for Thingy:91 to the GPS sample that doesn't use the driver
- Remove AT commands for COEX0 from the LTE link controller

jtguggedal commented:
Not sure what's going on with the failing compliance check, it works locally:
```
Undefined Kconfig symbols:

warning: undefined symbol BOARD_THINGY91_NRF9160NS:

- Referenced at /home/runner/work/sdk-nrf/sdk-nrf/ncs/nrf/drivers/gps/nrf9160_gps/Kconfig:73:

config NRF9160_GPS_SET_MAGPIO
	bool "Let the driver set MAGPIO configuration"
	default y if BOARD_NRF9160DK_NRF9160NS || BOARD_THINGY91_NRF9160NS
	depends on NRF9160_GPS_HANDLE_MODEM_CONFIGURATION && NRF9160_GPS
```

DurandA commented:
Sorry to hijack the PR but I have some questions about how COEX0 and MAGPIO are used.

* Does the LNA act as a passthru when COEX0 is high?
* How is MAGPIO used?
* For Thingy:91, is it ok to use an external antenna with this configuration:
  * passive antenna with LNA enabled 
  * active antenna with LNA disabled

![thingy-lna2](https://user-images.githubusercontent.com/3709037/95459345-d0913b80-0973-11eb-8241-0db891c8aa0e.png)

jtguggedal commented:
I'm wondering if the compliance check fails because the Kconfig option it complains about is defined in a downstream-only board file?

The CI failure is for an unrelated 52-board.

anangl commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -35,7 +35,17 @@
+#endif /* CONFIG_BOARD_NRF9160DK_NRF9160NS */
+
+#ifdef CONFIG_BOARD_THINGY91_NRF9160NS
+#define AT_MAGPIO      "AT\%XMAGPIO=1,1,1,7,1,746,803,2,698,748,2,1710,2200," \
+			"3,824,894,4,880,960,5,791,849,7,1565,1586"
+#ifdef CONFIG_GPS_SAMPLE_ANTENNA_ONBOARD
+#define AT_COEX0       "AT\%XCOEX0=1,1,1570,1580"
```
I think you need to update also [line 53](https://github.com/nrfconnect/sdk-nrf/pull/3077/files#diff-58f3c427fa4c71d51808275ce17d3f89R53) so that those macros are actually used.

jtguggedal replied:
Updated, that was a bad miss..

anangl commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -35,7 +35,17 @@
+#define AT_MAGPIO      "AT\%XMAGPIO=1,1,1,7,1,746,803,2,698,748,2,1710,2200," \
+			"3,824,894,4,880,960,5,791,849,7,1565,1586"
+#ifdef CONFIG_GPS_SAMPLE_ANTENNA_ONBOARD
+#define AT_COEX0       "AT\%XCOEX0=1,1,1570,1580"
```
Is the frequency range intentionally different here than in that Kconfig option default and `AT%MAGPIO` above? Seeing https://github.com/nrfconnect/sdk-nrf/commit/e522e4f426148e9355ac66c5e573d1fbfb6a3e24 I'm a bit in doubt.

jtguggedal replied:
Updated to the range that new modems are certified for

NordicBuilder commented:
**All checks passed.**

## checkpatch (informational only, not a failure)

```
No additional types will be considered - file &#x27;scripts/checkpatch/typedefsfile&#x27;: No such file or directory
-:142: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#142: FILE: samples/nrf9160/gps/src/main.c:54:
+    defined(CONFIG_BOARD_THINGY91_NRF9160NS)$

- total: 0 errors, 1 warnings, 122 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPDX_LICENSE_TAG SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.

```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

anangl commented on lib/lte_link_control/lte_lc.c:
```diff
@@ -184,19 +184,10 @@ static const char *const system_mode_params[] = {
-			"7,1565,1586",
-			"AT%XCOEX0=1,1,1565,1586"};
-#elif defined(CONFIG_BOARD_NRF9160DK_NRF9160NS)
-	const char *const lwm2m_ant_cfg[] = {
```
Since this is removed, the outdated line:
https://github.com/nrfconnect/sdk-nrf/blob/eb068b70e303f8c5f404833f160641575743c763/lib/lte_link_control/lte_lc.c#L553
should be also removed, to not generate confusion.

jtguggedal replied:
Done. Also did some other changes to improve readability.
The MAGPIO command will now be sent for Thingy:91 if `CONFIG_BSD_LIBRARY_SYS_INIT=n`, and not only if LwM2M carrier library is enabled. This catches more cases where this command is required in the link controller.

anangl commented on lib/lte_link_control/lte_lc.c:
```diff
@@ -176,27 +176,14 @@ static const char *const system_mode_params[] = {
-#endif
+#if !defined(CONFIG_BSD_LIBRARY_SYS_INIT) && \
+	defined(CONFIG_BOARD_THINGY91_NRF9160NS)
+const char thingy91_magpio[] = {
```
This should be `static`.

jtguggedal replied:
Indeed, fixed.

anangl commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -31,16 +31,27 @@
-static const char            update_indicator[] = {'\\', '|', '/', '-'};
-static const char            at_commands[][31]  = {
+static const char		update_indicator[] = {'\\', '|', '/', '-'};
+static const char		at_commands[][sizeof(AT_MAGPIO)]  = {
```
Why isn't it actually array of pointers? A few bytes in flash could be saved then.

jtguggedal replied:
Good point, the array size blew up quite a bit on Thingy:91 in particular. I'll do that change along with others later in the review, if there are any.

anangl commented:
> I'm wondering if the compliance check fails because the Kconfig option it complains about is defined in a downstream-only board file?

@jtguggedal It looks like it fails because of invalid dependencies of `BOARD_THINGY91_NRF9160*` options and a correction like this one: https://github.com/zephyrproject-rtos/zephyr/commit/34180d376f23e13487b8e78e327b5f68a225f1a4 is needed. I think it will be okay to add such commit here instead of creating a separate PR that this one will depend on.

jtguggedal commented:
> > I'm wondering if the compliance check fails because the Kconfig option it complains about is defined in a downstream-only board file?
> 
> @jtguggedal It looks like it fails because of invalid dependencies of `BOARD_THINGY91_NRF9160*` options and a correction like this one: [zephyrproject-rtos/zephyr@34180d3](https://github.com/zephyrproject-rtos/zephyr/commit/34180d376f23e13487b8e78e327b5f68a225f1a4) is needed. I think it will be okay to add such commit here instead of creating a separate PR that this one will depend on.

Thanks! I've added a commit for this, please let me know if it looks sane.
I also changed to using char pointer array for the AT commands in the GPS sample.

tejlmand commented on drivers/gps/nrf9160_gps/Kconfig:
```diff
@@ -72,26 +72,27 @@ endchoice
 config NRF9160_GPS_MAGPIO_STRING
 	string "MAGPIO string"
 	default "AT\%XMAGPIO=1,0,0,1,1,1565,1586" if BOARD_NRF9160DK_NRF9160NS
+	default "AT\%XMAGPIO=1,1,1,7,1,746,803,2,698,748,2,1710,2200,3,824,894,4,880,960,5,791,849,7,1565,1586" if BOARD_THINGY91_NRF9160NS
```
Kconfig settings like this has failed in SES menuconfig before.
Has it been verified that this work as expected when using latest SES-Nordic Edition 5.10c ?

jtguggedal replied:
Oh okay, I'll check and see how it works out with SES

jtguggedal replied:
Using SES 5.10c, the project loads, compiles and runs nicely on Thingy:91.

tejlmand replied:
thanks for verifying.

tejlmand commented on drivers/gps/nrf9160_gps/Kconfig:
```diff
@@ -72,26 +72,27 @@ endchoice
 	string "COEX0 string"
-	default "AT\%XCOEX0=1,1,1565,1586" if BOARD_NRF9160DK_NRF9160NS && NRF9160_GPS_ANTENNA_ONBOARD
-	default "AT\%XCOEX0" if BOARD_NRF9160DK_NRF9160NS && NRF9160_GPS_ANTENNA_EXTERNAL
+	default "AT\%XCOEX0=1,1,1565,1586" if (BOARD_NRF9160DK_NRF9160NS || BOARD_THINGY91_NRF9160NS) && NRF9160_GPS_ANTENNA_ONBOARD
+	default "AT\%XCOEX0" if (BOARD_NRF9160DK_NRF9160NS || BOARD_THINGY91_NRF9160NS) && NRF9160_GPS_ANTENNA_EXTERNAL
```
Kconfig settings like this has failed in SES menuconfig before.
Has it been verified that this work as expected when using latest SES-Nordic Edition 5.10c ?

tejlmand reviewed (commented):
A small question, nothing blocking.

jtguggedal commented:
CI passes, but the compliance check is still failing, and I've sort of run out of ideas on what to fix to make it pass.
Suspecting that it's because `thingy91_nrf9160ns` is a downstream-only board, and that those boards are possible not added to the board root in the compliance check scripts run on GitHub. Perhaps you have any ideas on the issue, @tejlmand ?

tejlmand commented:
> Suspecting that it's because `thingy91_nrf9160ns` is a downstream-only board, and that those boards are possible not added to the board root in the compliance check scripts run on GitHub.

That's correct, it's the compliance script here: https://github.com/nrfconnect/sdk-zephyr/blob/master/scripts/ci/check_compliance.py#L307
that seems to be a bit to fixed.
